object-construction-checker/tests/alwayscall/ACRegularExitPointTest.java

```diff
@@ -0,0 +1,274 @@
+            Foo fn2 = makeFoo();
+            fn2.a();
+            return fn2;
+        };
+
+        innerfunc.apply(f);
+
+    }
+
+
+    void ifElse(boolean b) {
+        if (b) {
+            Foo f1 = new Foo();
+            f1.a();
+        } else {
+            // :: error: missing.alwayscall
+            Foo f2 = new Foo();
+        }
+
+    }
+
+    Foo ifElseWithReturnExit(boolean b, boolean c) {
+        // :: error: missing.alwayscall
+        Foo f1 = makeFoo();
+        // :: error: missing.alwayscall
+        Foo f3 = new Foo();
+        // :: error: missing.alwayscall
+        Foo f4 = new Foo();
+
+        if (b) {
+            // :: error: missing.alwayscall
+            Foo f2 = new Foo();
+            if(c){
+                f4.a();
+            }else{
+                f4.b();
+            }
+            return f1;
+        }else {
+            // :: error: missing.alwayscall
+            Foo f2 = new Foo();
+            f2 = new Foo();
+            f2.a();
+        }
+        return f3;
+    }
+
+
+    void ifElseWithDeclaration(boolean b) {
+        Foo f1;
+        Foo f2;
+        if (b) {
+            f1 = new Foo();
+            f1.a();
+        } else {
+            // :: error: missing.alwayscall
+            f2 = new Foo();
+        }
+    }
+
+    void ifElseWithInitialization(boolean b) {
+        // :: error: missing.alwayscall
+        Foo f2 = new Foo();
+        Foo f11 = null;
+        if (b) {
+            f11 = makeFoo();
+            f11.a();
+        } else {
+            // :: error: missing.alwayscall
+            f2 = new Foo();
+        }
+    }
+
+    void ifWithInitialization(boolean b) {
+        // :: error: missing.alwayscall
+        Foo f1 = new Foo();
+        // :: error: missing.alwayscall
+        Foo f2 = new Foo();
+        if (b) {
+            f1.a();
+        }
+    }
+
+    void variableGoesOutOfScope(boolean b) {
+        if (b) {
+            Foo f1 = new Foo();
+            f1.a();
+        }
+    }
+
+
+    void ifWithNullInitialization(boolean b) {
+        Foo f1 = null;
+        Foo f2 = null;
+        if (b) {
+            f1 = new Foo();
+            f1.a();
+        } else {
+            // :: error: missing.alwayscall
+            f2 = new Foo();
+        }
+    }
+
+    void variableInitializedWithNull() { Foo f = null; }
+
+    void testLoop() {
+        Foo f = null;
+        while (true) {
+            // :: error: missing.alwayscall
+            f = new Foo();
+        }
+    }
+
+
+    void overWrittingVarInLoop() {
+        // :: error: missing.alwayscall
+        Foo f = new Foo();
+        while (true) {
+            // :: error: missing.alwayscall
+            f = new Foo();
+        }
+    }
+
+
+    void loopWithNestedBranches(boolean b) {
+        Foo f = null;
+        while (true) {
+            if (b) {
+                // :: error: missing.alwayscall
+                f = new Foo();
+            } else {
+                //TODO
```

Comment:
It relates to the second open problem in the design doc. If we have f.a() inside the else branch, the checker reports an error for "f = new Foo()" inside the else branch. It will be fixed in the next pull request.